x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/aggregate/TopTests.java

```diff
@@ -285,32 +450,75 @@ public static Iterable<Object[]> parameters() {
 
     @Override
     protected Expression build(Source source, List<Expression> args) {
-        return new Top(source, args.get(0), args.get(1), args.size() == 3 ? args.get(2) : null);
+        Expression field = args.get(0);
+        Expression outputField = args.size() > 3 ? args.get(3) : null;
+        if (field instanceof FieldAttribute f && outputField instanceof FieldAttribute of && f.fieldName().equals(of.fieldName())) {
+            // In order to avoid passing the same field twice as two different FieldAttribute objects, we use `field` as the fourth argument
+            // if `field`'s and `outputField`'s `fieldName` match. In such case TOP will use surrogate.
+            return new Top(source, field, args.get(1), args.size() > 2 ? args.get(2) : null, field);
+        } else {
+            return new Top(source, field, args.get(1), args.size() > 2 ? args.get(2) : null, outputField);
+        }
     }
 
     @SuppressWarnings("unchecked")
     private static TestCaseSupplier makeSupplier(
         TestCaseSupplier.TypedDataSupplier fieldSupplier,
         TestCaseSupplier.TypedDataSupplier limitCaseSupplier,
-        String order
+        String order,
+        TestCaseSupplier.TypedDataSupplier outputFieldSupplier
     ) {
         boolean isAscending = order == null || order.equalsIgnoreCase("asc");
-        boolean noOrderSupplied = order == null;
+        boolean orderSupplied = order != null;
+        boolean outputFieldSupplied = outputFieldSupplier != null;
+
+        List<DataType> dataTypes = new ArrayList<>();
+        dataTypes.add(fieldSupplier.type());
+        dataTypes.add(DataType.INTEGER);
+        if (orderSupplied) {
+            dataTypes.add(DataType.KEYWORD);
+        }
+        if (outputFieldSupplied) {
+            dataTypes.add(outputFieldSupplier.type());
+        }
 
-        List<DataType> dataTypes = noOrderSupplied
-            ? List.of(fieldSupplier.type(), DataType.INTEGER)
-            : List.of(fieldSupplier.type(), DataType.INTEGER, DataType.KEYWORD);
+        DataType expectedType = outputFieldSupplied ? outputFieldSupplier.type() : fieldSupplier.type();
 
         return new TestCaseSupplier(fieldSupplier.name(), dataTypes, () -> {
             var fieldTypedData = fieldSupplier.get();
             var limitTypedData = limitCaseSupplier.get().forceLiteral();
             var limit = (int) limitTypedData.getValue();
-            var expected = fieldTypedData.multiRowData()
-                .stream()
-                .map(v -> (Comparable<? super Comparable<?>>) v)
-                .sorted(isAscending ? Comparator.naturalOrder() : Comparator.reverseOrder())
-                .limit(limit)
-                .toList();
+            TestCaseSupplier.TypedData outputFieldTypedData;
+            List<?> expected;
+            if (outputFieldSupplied) {
```

Comment:
nit: To calculate the expected, maybe I would just always treat it as if the `outputParam == fieldParam` when `outputFieldSupplied == false`, and remove the old branch.

Take this with a grain of salt; I just feel like this method is getting quite chonky, but not a big deal as it's well localized 🚀

Reply:
> nit: To calculate the expected, maybe I would just always treat it as if the outputParam == fieldParam when outputFieldSupplied == false, and remove the old branch.

Done.